pyqtorch/noise.py

```diff
@@ -0,0 +1,420 @@
+                raise ValueError("The probability value is not a correct probability")
+        # Verification list probability values:
+        elif isinstance(self.probability, list):
+            sum_prob = sum(self.probability)
+            if sum_prob > 1.0:
+                raise ValueError("The sum of probabilities can't be greater 1.0")
+            for proba in self.probability:
+                if proba > 1.0 or proba < 0.0:
+                    raise ValueError("The probability values are not correct probabilities")
+
+    @property
+    def proba(self) -> Union[Tuple[float, ...], float]:
+        return self.probability
+
+    @property
+    def name(self) -> str:
+        return self.__class__.__name__
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, type(self)):
+            return self.name == other.name and self.probabilty == other.probabilty
+        else:
+            return False
+
+    def extra_repr(self) -> str:
+        return f"'qubit_support'={self.qubit_support}, 'probability'={self.probability}"
+
+    @property
+    def kraus_operator(self) -> List[Tensor]:
+        return self.kraus
+
+    # Since PyQ expects tensor.Size = [2**n_qubits, 2**n_qubits,batch_size].
+    def unitary(self) -> List[Tensor]:
+        return [kraus_op.unsqueeze(2) for kraus_op in self.kraus]
+
+    def dagger(self) -> List[Tensor]:
+        return [_dagger(kraus_op) for kraus_op in self.unitary()]
+
+    def forward(self, state: Tensor) -> Tensor:
+        """
+        Applies a noisy quantum channel on the input state.
+        The evolution is represented as a sum of Kraus operators:
+        .. math::
+            S(\\rho) = \\sum_i K_i \\rho K_i^\\dagger,
+
+        Each Kraus operator in the `kraus_list` is applied to the input state, and the result
+        is accumulated to obtain the evolved state.
+
+        Args:
+            state (Tensor): Input quantum state represented as a tensor.
+
+        Returns:
+            Tensor: Quantum state as a density matrix after evolution.
+
+        Raises:
+            TypeError: If the input `state` or `kraus_list` is not a Tensor.
+        """
+
+        # Verification input type:
+        if not isinstance(state, Tensor):
+            raise TypeError("The input must be a Tensor")
+
+        # Output operator initialization:
+        n_qubits: int = len(state.size()) - 1
+        batch_size: int = state.size(-1)
+        rho_evol: Tensor = torch.zeros(
+            2**n_qubits, 2**n_qubits, batch_size, dtype=DEFAULT_MATRIX_DTYPE
+        )
+
+        # Apply noisy channel on input state
+        rho: Tensor = density_mat(state)
+        kraus_unit: List[Tensor] = self.unitary()
+        kraus_dag: List[Tensor] = self.dagger()
+        for i in range(len(self.kraus)):
+            rho_i: Tensor = apply_ope_ope(
+                kraus_unit[i], apply_ope_ope(rho, kraus_dag[i], self.target), self.target
+            )
+            rho_evol += rho_i
```

Comment:
Maybe stack the tensors ?

Reply:
I not sure to get it. What's your idea ? What do you to stack and why?

Reply:
Well, instead of initialising `rho_evol` as a tensor of zeros and adding on top of that, mayeb use `torch.stack`:

```python
A = torch.stack(A, B) <-- here you stack tensor B to A
```
To be confirmed though.

Reply:
Here I am not stacking tensors but performing addition in the idea to compute  $\sum_i K_i \rho K^{\dagger}_i $, right ?
Also `torch.stack` take as input a `tuple[Tensor,...]`: https://pytorch.org/docs/stable/generated/torch.stack.html

Reply:
Does this version fit you ?
```python3
    rho_evol = torch.stack(
            [
                operator_product(
                    kraus_unitary, operator_product(rho, kraus_dagger, self.target), self.target
                )
                for kraus_unitary, kraus_dagger in zip(self.unitary(), self.dagger())
            ],
            dim=0,
        )
        rho_evol = torch.sum(rho_evol, dim=0)
        return rho_evol